### PR TITLE
feat: Implement bundle block range filtering

### PR DESCRIPTION
**Pull request by nicholaspai**
- Implements `getNextBundleStartBlockNumber` in HubPoolClient which can be used by data worker to construct bundle block range for new root bundle proposal
- Adds unit tests for above function
- Implements first part of `Dataworker.proposeRootBundle` which constructs block range for next root using `getNextBundleStartBlockNumber` and then inputs that block range into root builder functions to construct root for range

Still todo:
- add tests for `Dataworker` methods that check that block range filtering is used correctly

### CODE REVIEW

**Comment by nicholaspai** [comment]
Will add some tests after first pass on reviews

File: src/clients/HubPoolClient.ts [hunk 1]
```diff
@@ -113,6 +114,33 @@ export class HubPoolClient {
     return endingBlockNumber;
   }
 
+  getNextBundleStartBlockNumber(chainIdList: number[], latestBlock: number, chainId: number): number {
```
Comment (chrismaree) [inline src/clients/HubPoolClient.ts:117]: so your PR works by using the `spreadEventWithBlockNumber` which appends the block number to the event then uses `getNextBundleStartBlockNumber` to do the searching over the provided range. this makes sence. we can think about refining this later to contain ranges and searching within the underlying class structure. what you have for now does make sence.

File: src/clients/HubPoolClient.ts [hunk 2]
```diff
@@ -113,6 +114,33 @@ export class HubPoolClient {
...
+      (event: ExecutedRootBundle) => event.blockNumber <= latestBlock && event.chainId === chainId
+    );
+
+    // TODO: If no event for chain ID, then we can return a conservative default starting block like 0,
+    // or we could throw an Error.
+    if (!latestRootBundleExecutedEvent) return 0;
+
+    // Once that event is found, search for the ProposeRootBundle event that is as late as possible, but earlier than
+    // the RootBundleExecuted event we just identified.
+    const mostRecentProposeRootBundleEventWithChain = sortEventsDescending(this.proposedRootBundles).find(
```
Comment (chrismaree) [inline src/clients/HubPoolClient.ts:130]: in what situation would the events not be sorted?

**Review by chrismaree — APPROVED** [review]
this LGTM. I think we still need to think about the overall data structures we are using in the HubPoolClient but for now this is ok as a path of least resistance.

**Pull request merged**